packages/react-dom/src/__tests__/ReactDOMServerIntegrationElements-test.js

```diff
@@ -860,17 +876,23 @@ describe('ReactDOMServerIntegration', () => {
               {'\r\nbaz\nqux\u0000'}
             </div>,
           );
-          if (render === serverRender || render === streamRender) {
+          if (
+            render === serverRender ||
+            render === streamRender ||
+            (render === clientRenderOnServerString &&
+              ReactFeatureFlags.enableClientRenderFallbackOnTextMismatch)
+          ) {
             // We have three nodes because there is a comment between them.
             expect(e.childNodes.length).toBe(3);
-            // Everything becomes LF when parsed from server HTML.
+            // Everything becomes LF when parsed from server HTML or hydrated if enableClientRenderFallbackOnTextMismatch is on.
             // Null character is ignored.
             expectNode(e.childNodes[0], TEXT_NODE_TYPE, 'foo\nbar');
             expectNode(e.childNodes[2], TEXT_NODE_TYPE, '\nbaz\nqux');
           } else if (render === clientRenderOnServerString) {
             // We have three nodes because there is a comment between them.
             expect(e.childNodes.length).toBe(3);
-            // Hydration uses JS value with CR and null character.
+            // Hydration without enableClientRenderFallbackOnTextMismatch uses JS value with CR and null character.
```

Comment:
Hydrating with the new root without `enableClientRenderFallbackOnTextMismatch` (www) is like old root i.e. replaces server HTML with client side rendered CR.
Hydrating with the new root with `enableClientRenderFallbackOnTextMismatch` is keeping the server HTML.

Looks like not patching up the server HTML here was intentionally added in https://github.com/facebook/react/pull/26391

Reply:
Nice explanation, thanks for the fix!